File: packages/replay-web/src/size.ts
import { GameSize, DeviceSize } from "@replay/core";
import { Dimensions } from "./dimensions";

export function calculateDeviceSize(
  innerWidth: number,
  innerHeight: number,
  dimensions: Dimensions,
  size: GameSize
): DeviceSize {
  let orientationSize;
  if ("portrait" in size) {
    const isPortrait = innerHeight > innerWidth;
    orientationSize = isPortrait ? size.portrait : size.landscape;
  } else {
    orientationSize = size;
  }
  const {
    width,
    height,
    maxWidthMargin = 0,
    maxHeightMargin = 0,
  } = orientationSize;

  if (
    dimensions === "game-coords" ||
    // avoid dividing by 0s
    innerWidth === 0 ||
    innerHeight === 0
  ) {
    return {
      width,
      height,
      widthMargin: 0,
      heightMargin: 0,
      fullWidth: width,
      fullHeight: height,
      deviceWidth: width,
      deviceHeight: height,
    };
  }

  // scale the game up into browser window
  const windowWidthHeightRatio = innerWidth / innerHeight;
  const gameWidthHeightRatio = width / height;

  if (gameWidthHeightRatio > windowWidthHeightRatio) {
    // game is wider than window, so set width to match window and scale height with margin
    const scaledWidth = innerWidth;
    const noMarginScaledHeight = scaledWidth / gameWidthHeightRatio;
    const scaledMaxHeightMargin =
      (noMarginScaledHeight / height) * maxHeightMargin;
    const scaledHeight = Math.min(
      innerHeight,
      noMarginScaledHeight + scaledMaxHeightMargin * 2
    );
    const heightMargin =
      ((scaledHeight - noMarginScaledHeight) / 2) *
      (height / noMarginScaledHeight);

    return {
      width,
      height,
      widthMargin: 0,
      // height margin in px, then convert to game coordinates
      heightMargin,
      fullWidth: width,
      fullHeight: height + heightMargin * 2,
      deviceWidth: scaledWidth,
      deviceHeight: scaledHeight,
    };
  } else {
    // game is taller than window, so set height to match window and scale width
    // with margin
    const scaledHeight = innerHeight;
    const noMarginScaledWidth = scaledHeight * gameWidthHeightRatio;
    const scaledMaxWidthMargin = (noMarginScaledWidth / width) * maxWidthMargin;
    const scaledWidth = Math.min(
      innerWidth,
      noMarginScaledWidth + scaledMaxWidthMargin * 2
    );
    const widthMargin =
      ((scaledWidth - noMarginScaledWidth) / 2) * (width / noMarginScaledWidth);

    return {
      width,
      height,
      widthMargin,
      heightMargin: 0,
      fullWidth: width + widthMargin * 2,
      fullHeight: height,
      deviceWidth: scaledWidth,
      deviceHeight: scaledHeight,
    };
  }
}

File: packages/replay-swift/Replay/Sources/Replay/Size.swift
import UIKit

struct Size {
    // Which of these is used is determined by user's Xcode project settings
    var portrait: OrientationSize
    var landscape: OrientationSize
}

struct OrientationSize {
    var width: CGFloat
    var height: CGFloat

    var minWidthXL: CGFloat?
    var minHeightXL: CGFloat?

    // default 0
    var maxWidthMargin: CGFloat
    var maxHeightMargin: CGFloat
}

struct GameViewSize: Equatable {
    var gameSize: CGSize
    var offset: CGPoint
    var scale: CGFloat
}

class SizeUtils {
    static func calculateDeviceSize(frame viewSize: CGRect, gameSize: OrientationSize) -> (DeviceSize, GameViewSize) {
        let viewWidthHeightRatio = viewSize.width / viewSize.height;
        let gameWidthHeightRatio = gameSize.width / gameSize.height;

        if (gameWidthHeightRatio > viewWidthHeightRatio) {
            // game is wider than view, so set width to match view and scale height with margin
            let scaledWidth = viewSize.width;
            let noMarginScaledHeight = scaledWidth / gameWidthHeightRatio;
            let scaledMaxHeightMargin =
                (noMarginScaledHeight / gameSize.height) * gameSize.maxHeightMargin;
            let scaledHeight = min(
                viewSize.height,
                noMarginScaledHeight + scaledMaxHeightMargin * 2
            );
            let heightMargin = ((scaledHeight - noMarginScaledHeight) / 2) * (gameSize.height / noMarginScaledHeight)
            let totalGameHeight = gameSize.height + heightMargin * 2
            return (
                DeviceSize(
                    width: cgFloatToNsNumber(gameSize.width),
                    height: cgFloatToNsNumber(gameSize.height),
                    widthMargin: 0,
                    heightMargin: cgFloatToNsNumber(heightMargin),
                    deviceWidth: cgFloatToNsNumber(scaledWidth),
                    deviceHeight: cgFloatToNsNumber(scaledHeight)
                ),
                GameViewSize(
                    gameSize: CGSize(width: gameSize.width, height: totalGameHeight),
                    offset: CGPoint(x: (viewSize.width - gameSize.width) / 2, y: (viewSize.height - totalGameHeight) / 2),
                    scale: viewSize.width / gameSize.width
                )
            )
        } else {
            // game is taller than view, so set height to match view and scale width with margin
            let scaledHeight = viewSize.height;
            let noMarginScaledWidth = scaledHeight * gameWidthHeightRatio;
            let scaledMaxWidthMargin = (noMarginScaledWidth / gameSize.width) * gameSize.maxWidthMargin;
            let scaledWidth = min(
                viewSize.width,
                noMarginScaledWidth + scaledMaxWidthMargin * 2
            );
            let widthMargin = ((scaledWidth - noMarginScaledWidth) / 2) * (gameSize.width / noMarginScaledWidth)
            let totalGameWidth = gameSize.width + widthMargin * 2
            return (
                DeviceSize(
                    width: cgFloatToNsNumber(gameSize.width),
                    height: cgFloatToNsNumber(gameSize.height),
                    widthMargin: cgFloatToNsNumber(widthMargin),
                    heightMargin: 0,
                    deviceWidth: cgFloatToNsNumber(scaledWidth),
                    deviceHeight: cgFloatToNsNumber(scaledHeight)
                ),
                GameViewSize(
                    gameSize: CGSize(width: totalGameWidth, height: gameSize.height),
                    offset: CGPoint(x: (viewSize.width - totalGameWidth) / 2, y: (viewSize.height - gameSize.height) / 2),
                    scale: viewSize.height / gameSize.height
                )
            )
        }
    }

    static func deviceXToGameX(x: CGFloat, deviceSize: DeviceSize) -> CGFloat {
        let width = CGFloat(truncating: deviceSize.width)
        let widthMargin = CGFloat(truncating: deviceSize.widthMargin)
        return x - width / 2 - widthMargin
    }
    static func gameXToDeviceX(x: NSNumber, deviceSize: DeviceSize) -> CGFloat {
        let width = CGFloat(truncating: deviceSize.width)
        let widthMargin = CGFloat(truncating: deviceSize.widthMargin)
        return CGFloat(truncating: x) + width / 2 + widthMargin
    }

    static func deviceYToGameY(y: CGFloat, deviceSize: DeviceSize) -> CGFloat {
        let height = CGFloat(truncating: deviceSize.height)
        let heightMargin = CGFloat(truncating: deviceSize.heightMargin)
        return height / 2 + heightMargin - y
    }
    static func gameYToDeviceY(y: NSNumber, deviceSize: DeviceSize) -> CGFloat {
        let height = CGFloat(truncating: deviceSize.height)
        let heightMargin = CGFloat(truncating: deviceSize.heightMargin)
        return height / 2 + heightMargin - CGFloat(truncating: y)
    }
}
